proposals/0000-local-modules.rst

```diff
@@ -0,0 +1,1330 @@
+   thus the names ``Zero`` and ``Succ`` apply well to each. This is
+   impossible today, as the names clash.
+
+   With this proposal's optional change 4.3, these declarations would be accepted. The constructors
+   would be disambiguated with module prefixes, like ``Nat.Zero`` and
+   ``Elem.Succ``. (The ``Fin`` declaration would need to say ``Nat.Succ`` in
+   place of ``Succ``.) An unqualified use of a constructor would be an error.
+   Alternatively, to prevent ``Fin``\'s and ``Elem``\'s constructors from
+   being available unqualified in the global scope, the user could declare
+   ``data qualified Fin ...`` and
+   ``data qualified Elem ...``.
+
+3. When a function ``f`` needs a helper ``h``, we can declare ``h`` in a
+   ``where`` clause. However, suppose both ``f`` and ``g`` need ``h``. Now,
+   ``h`` must be declared at the top level, meaning all the rest of the
+   definitions in the module can see ``h``, even if ``h`` is really very
+   specific to ``f`` and ``g``.
+
+   With this proposal, we can model this situation nicely::
+
+     module (f, g) where
+       f :: ...
+       f = ...
+
+       g :: ...
+       g = ...
+
+       h :: ...
+       h = ...
+
+   Unfortunately, there is no way for ``h`` to access arguments passed to
+   ``f`` or ``g`` without declaring these as arguments to ``h`` and passing
+   them explicitly. Fixing this was the subject of `#40`_; see `Future Work`_.
+
+4. If a function or group of functions needs to work with a datatype or class
+   locally, there is no way to do this without polluting the namespace of the
+   entire file.
+
+   With this proposal, we can do this easily::
+
+     module (f) where
+       f :: ...
+       f = ...
+
+       data SpecialDataType = ...
+
+       class LocalClass a b c where ...
+
+5. When we expect users to import our library qualified, we have no way of
+   signaling this beyond documentation; we also have no way of making it easy
+   to import the module qualified correctly.
+
+   With this proposal, we can do this easily::
+
+     -- top of file:
+     module Data.Set ( Set, module qualified Set ) where
+
+       module Set ( Set, fromList ) where
+         data Set = ...
+         fromList = ...
+
+   An importer who days ``import Data.Set`` will get access to ``Set`` (the
+   type) and ``Set.fromList``, the function. The fact that the module and type
+   have the same name is inconsequential here, but it is permitted.
+
+6. Suppose our library exports class ``C`` with method ``meth``. Our users
+   will frequently import ``C(..)`` and get ``meth`` in scope. Now I wish to
+   refactor ``C`` without changing my users' import behavior. There is no way
+   to do this currently.
+
+   With this proposal, we can do this::
+
+     -- top of file:
+     module MyLibrary ( module C(meth) ) where
+
+       class C a where ...
+
+       module C where
+         meth :: ...
+
+   This example shows that modules may be *extended*. The ``class C``
```

Comment:
Extensions are always allowed, everywhere. This is because a "module" isn't really an entity -- it's just a namespace... which is really saying that it's just a prefix used in qualified names.

I suppose this could be taken advantage of. For instance if you have a local module `Trusted` exported from some trusted, anyone could inject more definitions into that namespace. But if any of these definitions share a name with something in the original `Trusted` namespace, usages become ambiguous. So you'd have to also somehow trick the user into a using a function not listed in the documentation for `Trusted`. If you're importing a library with such ill-intent, you've got bigger problems than just namespace injection. I'm not bothered by this scenario.